src/main/java/com/ibm/etcd/client/utils/RangeCache.java

```diff
@@ -210,22 +208,40 @@ public RangeCache(EtcdClient client, ByteString fromKey, ByteString toKey,
                         .map(ResponseOp::getResponseRange)
                         .collect(Collectors.toList()), directExecutor());
         }
-        
-        final SettableFuture<Boolean> promise = new SettableFuture<Boolean>() {
+
+        final class StartPromise extends SettableFuture<Boolean> implements Runnable {
+            volatile ListenableFuture<?> parent;
+            StartPromise(ListenableFuture<?> parent) {
+                this.parent = parent;
+                parent.addListener(this, directExecutor());
+            }
             @Override
-            protected void interruptTask() {
-                if (rrfut.cancel(true)) {
+            public void run() {
+                parent = null; // runs when parent completes
+            }
+            @Override
+            protected void afterDone() {
+                if (!wasInterrupted()) {
+                    return;
+                }
+                // propagate cancellation
+                final ListenableFuture<?> p = parent;
+                if (p != null && p.cancel(true)) {
                     return;
                 }
                 Watch theWatch;
                 synchronized(RangeCache.this) {
                     theWatch = watch;
+                    closed = true;
```

Comment:
Was this another bug where canceling the range request wouldn't end up with `closed` set?

Reply:
Possibly, though I think pretty benign, don't think it would even had any detrimental effect.